src/nvim/change.c

```diff
@@ -518,6 +526,13 @@ void deleted_lines_mark(linenr_T lnum, int count)
 /// @param xtra number of extra lines (negative when deleting)
 void changed_lines_redraw_buf(buf_T *buf, linenr_T lnum, linenr_T lnume, linenr_T xtra)
 {
+  // If lines haven been inserted/deleted and the buffer has virt_lines,
+  // mark the line after the changed lines to be redrawn as some virt_lines may
+  // now be drawn above a different line.
+  if (xtra != 0 && buf_meta_total(buf, kMTMetaLines) > 0) {
+    lnume++;
+  }
+
```

Comment:
This change actually reduces redraw when deleting a line with virtual lines attached, because it makes the range here the same as `changed_lines_invalidate_buf()`, which is a situation that `win_update()` handles better.